src/stem-tracker.js

```diff
@@ -0,0 +1,77 @@
+'use strict';
+var Stem = require("./stem");
+
+/**
+ * Stem Tracking Utility.
+ *
+ * Handles the stems for a given editor.
+ */
+module.exports = function StemTracker(editor, onClick) {
+  var me = this;
+  var previousStems = [];
+  var stems = [];
+  var topLevelTags = ['p', 'h1', 'h2', 'h3', 'div', 'section'];
+
+  // Listen for changes in content and update.
+  editor.addListener({
+    onContent: function () {
+      updateStems( editor.currentElem() );
+      return false;
+    },
+    onKey: function(e) {
+      if (e.keyType == 'backspace' && e.point.type == 'start') {
+        updateStems( editor.currentElem() );
+      }
+      return false; 
+    }
+  });
+
+  function updateStems(editorElem){
+    previousStems = stems;
+    stems = [];
+
+    
+    // Loop through all top level elements
+    var topLevelElems = getTopLevelElems( topLevelTags, editorElem );
+    for (var i=0; i < topLevelElems.length; i++){
+      var elem = topLevelElems[i];
+      // Add stems where required.
+      if (needsStem(elem)) {
+        var stem = Stem.getOrCreate( elem, onClick );
+        stems.push(stem);
+      }
+    }
+
+    // Remove orphaned stems
+    for (var i=0; i < previousStems.length; i++){
+      var previousStem = previousStems[i];
+      if (stems.indexOf(previousStem) < 0){
+        previousStem.destroy();
+      }
+    }
+
+
+  }
+
+  function getTopLevelElems(tags, editorElem){
+    var topLevelElems = [];
+    for (var i=0; i < tags.length; i++){
+      topLevelElems = mergeNodeList( 
+        topLevelElems, editorElem.getElementsByTagName( tags[i] ) 
```

Comment:
this would get all elements, even any random divs deep within an embedded widget.

a better way would be to simply loop over the editor's top level elements - in practice, these would in fact constitute all the nodes you care about - but you can also check with util.isBlock() to see if they're a block (instead of some junk text node), and we can add other checks later to exclude inter-paragraph widgets, depending on how those are implemented.  In short, something like

for (var node = editorElem.firstChild; node != null; node = node.nextSibling) {
  if (needsStem(node)) {
    ...
  }
}

then, augment needsStem as follows:

function needsStem(elem) {
  if (!canEverHaveStem(elem)) { 
    return false;
  }

  return ... current code you have ...
}

function canEverHaveStem(elem) {
  ... whatever logic makes sense, e.g. ...
  return util.isBlock(elem) && util.isEditable(elem); // isEditable would be basic first-cut logic to ignore widgets we've placed
}